plumbing/reference_validation.go

```diff
@@ -0,0 +1,311 @@
+	ErrRefOnlyAtSign                = errors.New("ref name cannot only be an at sign")
+)
+
+var (
+	PatternLeadingDot                  = regexp.MustCompile(`^\.`)
+	PatternTrailingLock                = regexp.MustCompile(`\.lock$`)
+	PatternAtLeastOneForwardSlash      = regexp.MustCompile(`^[^/]+$`)
+	PatternDoubleDots                  = regexp.MustCompile(`\.\.`)
+	PatternExcludedCharacters          = regexp.MustCompile(`[\000-\037\177 ~^:?*[]+`)
+	PatternLeadingForwardSlash         = regexp.MustCompile(`^/`)
+	PatternTrailingForwardSlash        = regexp.MustCompile(`/$`)
+	PatternConsecutiveForwardSlashes   = regexp.MustCompile(`//+`)
+	PatternTrailingDot                 = regexp.MustCompile(`\.$`)
+	PatternAtOpenBrace                 = regexp.MustCompile(`@{`)
+	PatternExcludedCharactersAlternate = regexp.MustCompile(`[\000-\037\177 ~^:?[]+`)
+	PatternOneAllowedAsterisk          = regexp.MustCompile(`^[^*]+?\*?[^*]+?$`)
+	PatternOnlyAtSign                  = regexp.MustCompile(`^@$`)
+)
+
+type CheckRefOptions struct {
+	// They must contain at least one /
+	//  If the --allow-onelevel option is used, this rule is waived.
+	AllowOneLevel bool
+	// If this option is enabled, <refname> is allowed to contain a
+	// single * in the refspec
+	RefSpecPattern bool
+	// Normalize refname by removing any leading slash (/) characters and
+	// collapsing runs of adjacent slashes between name components into
+	// a single slash.
+	Normalize bool
+}
+
+type ActionOptions struct {
+
+	// no slash-separated component can begin with a dot .
+	HandleLeadingDot ActionChoice
+
+	// no slash-separated component can end with the sequence .lock
+	HandleTrailingLock ActionChoice
+	// They must contain at least one /.
+	HandleAtLeastOneForwardSlash ActionChoice
+	// They cannot have two consecutive dots .. anywhere.
+	HandleDoubleDots ActionChoice
+	// They cannot have ASCII control characters (i.e. bytes whose values
+	// are lower than \040, or \177 DEL), space, tilde ~, caret ^, or
+	// colon : anywhere.
+	// They cannot have question-mark ?, asterisk *, or open
+	// bracket [ anywhere
+	// They cannot contain a \
+	HandleExcludedCharacters ActionChoice
+	// They cannot begin or end with a slash /
+	HandleLeadingForwardSlash ActionChoice
+	// They cannot begin or end with a slash / or contain
+	// multiple consecutive slashes
+	HandleTrailingForwardSlash ActionChoice
+	// They cannot  contain multiple consecutive slashes
+	HandleConsecutiveForwardSlashes ActionChoice
+	// They cannot end with a dot .
+	HandleTrailingDot ActionChoice
+	// They cannot contain a sequence @{
+	HandleAtOpenBrace ActionChoice
+	// They cannot be the single character @
+	HandleOnlyAtSign ActionChoice
+}
+
+// https://git-scm.com/docs/git-check-ref-format
+// git-check-ref-format
+type RefNameChecker struct {
+	Name            ReferenceName
+	CheckRefOptions CheckRefOptions
+	ActionOptions   ActionOptions
+}
+
+func NewCheckRefOptions(default_value bool) *CheckRefOptions {
```

Comment:
`s/default_value/default` and please read https://golang.org/doc/effective_go.html#mixed-caps